nixos/modules/installer/cd-dvd/sd-image-raspberrypi4.nix

```diff
@@ -18,14 +18,22 @@
 
   sdImage = {
     firmwareSize = 128;
-    # This is a hack to avoid replicating config.txt from boot.loader.raspberryPi
-    populateFirmwareCommands =
-      "${config.system.build.installBootLoader} ${config.system.build.toplevel} -d ./firmware";
-    # As the boot process is done entirely in the firmware partition.
-    populateRootCommands = "";
+    # /var/empty is needed for some services, such as sshd
+    populateRootCommands = "mkdir -p ./files/var/empty";
```

Comment:
If this is required, this shouldn't be handled here. This should be fixed upstream in NixOS so that when it is missing, it works. I think this explains an issue I saw on another board, but did not investigate.

At the very least, if it's not part of some tmpfiles declaration, it should be part of the first boot commands in the generic sd image builder.

Reply:
I thought this would be handled in the activation script:

https://github.com/NixOS/nixpkgs/blob/staging/nixos/modules/system/activation/activation-script.nix#L184-L198

Reply:
iirc Pi images are treated as if they're live CDs, which doesn't run this. I'm not sure though, I just know that without those commands, there wouldn't be a `/var/empty`.

Reply:
Yeah that sounds right, might need this for the other rpi images too.

Reply:
yes, I've used this same for rpi3:
```
                systemd.tmpfiles.rules = [
                    "d /var/empty - - - -"
                ];
 ```

Reply:
I wonder what's the failure mode here. Why isn't it required for other systems than the ARM images?

Reply:
On rpi3 I'm also experiencing sshd crashing with `Missing privilege separation directory: /var/empty`. Only on the first boot. If i reboot once, it's fine.
On the first boot when it's not working `dmesg | grep /var/empty` gives me:
```
[    8.485396] stage-2-init: mkdir: cannot create directory '/var/empty': No space left on device
[    8.536393] stage-2-init: find: '/var/empty': No such file or directory
[    8.545242] stage-2-init: chmod: cannot access '/var/empty': No such file or directory
[    8.556068] stage-2-init: chown: cannot access '/var/empty': No such file or directory
```
On the second boot there is nothing in dmesg related to `/var/empty`.
Should i open a bug report for this?

Reply:
Oh, that could be the root cause of this issue I guess. I would rather this gets fixed appropriately rather than patched over.

If you open a bug, it is not likely sshd-specific, but really this one issue. And please do, cc me once you do.

Reply:
I vote for merging this immediately, since a hack that fixes broken sshd is better than broken sshd.

Reply:
t. person that has wasted the last hours trying to log into sshd on his raspberrypi.